### PR TITLE
[Feature] Table function table() supports orc file

### LINKED ISSUE #21159: Simply load files from S3/HDFS
## Feature request

Now, we could load files from S3/HDFS by broker load and external file table.
### Broker Load
```
LOAD LABEL db0.label2 (     DATA INFILE("s3a://starrocks-common/fuwei/student.parquet")     INTO TABLE tbl_student     FORMAT AS "parquet"     (name, age)  ) WITH BROKER (     "aws.s3.access_key" = "xxx",     "aws.s3.secret_key" = "xxx",    "aws.s3.endpoint"="xxx.amazonaws.com" );
```
### External File Table 
```
CREATE EXTERNAL TABLE tbl_student_s3 (     name string,      id int )  ENGINE=file PROPERTIES  (     "path" = "xxx.parquet",      "format" = "parquet",      "aws.s3.access_key" = "xxx",     "aws.s3.secret_key" = "xxx",     "aws.s3.region" = "us-west-2"  );
```

These methods are not easy to use. Moreover, we need to know the schema of files before we load data.

**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by rickif**
Fixes #21159
This PR adds the support of reading orc file to `table()` table function, including following change:
1. support orc file schema inferring.

Here's an example how to use the `table()` to reads orc file from S3.
```
MySQL [db0]> select * from table(
    "path" = "s3a://xxx/scalar_types2.orc", 
    "format" = "orc",   
    "aws.s3.access_key" = "xxx",     
    "aws.s3.secret_key" = "xxx",     
    "aws.s3.region" = "us-west-2"
)\G;

*************************** 1. row ***************************
     col_bool: 1
  col_tinyint: 127
 col_smallint: 32767
      col_int: 2147483647
   col_bigint: 9223372036854775807
    col_float: 12335.6
   col_double: 1234567.8
   col_string: jack
     col_char: rose
  col_varchar: titanic
  col_decimal: 12345678.12
col_timestamp: 2011-11-03 20:05:23
     col_date: 2019-12-04
1 row in set (2.764 sec)

```

## What type of PR is this:
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [x] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 3 / 11 (27.27%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/TableFunctionTable.java|3|11|27.27%|[130, 167, 168, 171, 173, 174, 175, 180]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 1 / 8 (12.50%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/TableFunctionTable.java|1|8|12.50%|[178, 180, 181, 183, 184, 186, 192]|

**Review by chaoyli — DISMISSED** [review]

**Review by meegoo — DISMISSED** [review]

File: be/src/formats/orc/orc_chunk_reader.cpp [hunk 1]
```diff
@@ -1250,4 +1264,87 @@ bool OrcChunkReader::is_implicit_castable(TypeDescriptor& starrocks_type, const
...
+    auto const& root = _reader->getType();
+
+    auto cnt = root.getSubtypeCount();
+    for (uint64_t i = 0; i < cnt; i++) {
+        TypeDescriptor tp;
+        auto name = root.getFieldName(i);
+
+        auto subtype = root.getSubtype(i);
+        switch (subtype->getKind()) {
+        case orc::TypeKind::BOOLEAN:
```
Comment (mofeiatwork) [inline be/src/formats/orc/orc_chunk_reader.cpp:1277]: replace with a map ?

Comment (rickif) [reply be/src/formats/orc/orc_chunk_reader.cpp:1277]: I don't find a good way to use a map to handle types like `orc::TypeKind::STRING`. The SR column type is created by a function call `TypeDescriptor::create_varchar_type(TypeDescriptor::MAX_VARCHAR_LENGTH)`.

Comment (rickif) [reply be/src/formats/orc/orc_chunk_reader.cpp:1277]: I don't find a good way to use a map to handle types like `orc::TypeKind::STRING`. The SR column type is created by a function call `TypeDescriptor::create_varchar_type(TypeDescriptor::MAX_VARCHAR_LENGTH)`.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 1 / 8 (12.50%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/TableFunctionTable.java|1|8|12.50%|[178, 180, 181, 183, 184, 186, 192]|

**Review by murphyatwork — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 1 / 8 (12.50%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/TableFunctionTable.java|1|8|12.50%|[178, 180, 181, 183, 184, 186, 192]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 1 / 8 (12.50%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/TableFunctionTable.java|1|8|12.50%|[178, 180, 181, 183, 184, 186, 192]|

**Review by meegoo — APPROVED** [review]

**Review by chaoyli — APPROVED** [review]

**Comment by rickif** [comment]
@mergifyio backport branch-3.1

**Pull request merged**